### PR TITLE
Add Inject instances for Eons. Deprecate old eons conversion functions.

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Deprecate eons conversion functions like `conwayEraOnwardsToBabbageEraOnwards`. Add [`Inject`](https://cardano-ledger.cardano.intersectmbo.org/cardano-ledger-core/Cardano-Ledger-BaseTypes.html#t:Inject) instances for eon conversions. See the PR description for migration aid.
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
   - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
   - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Use [Inject](https://cardano-ledger.cardano.intersectmbo.org/cardano-ledger-core/Cardano-Ledger-BaseTypes.html#t:Inject) for eons conversion instead of specialised and hard to remember functions.

You can migrate your codebase using this snippet:
```bash
find . -type f -name "*.hs" -exec sed -i \
  -e 's/allegraEraOnwardsToShelleyBasedEra/inject/g' \
  -e 's/alonzoEraOnwardsToShelleyBasedEra/inject/g' \
  -e 's/babbageEraOnwardsToShelleyBasedEra/inject/g' \
  -e 's/conwayEraOnwardsToShelleyBasedEra/inject/g' \
  -e 's/conwayEraOnwardsToBabbageEraOnwards/inject/g' \
  -e 's/maryEraOnwardsToShelleyBasedEra/inject/g' \
  -e 's/shelleyEraOnlyToShelleyBasedEra/inject/g' \
  -e 's/shelleyToAllegraEraToShelleyBasedEra/inject/g' \
  -e 's/shelleyToBabbageEraToShelleyBasedEra/inject/g' \
  -e 's/shelleyToMaryEraToShelleyBasedEra/inject/g' \
  -e 's/babbageEraOnwardsToEra/inject/g' \
  -e 's/eraToBabbageEraOnwards/inject/g' {} +

```

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/gen/Test/Gen/Cardano/Api/Typed.hs [hunk 1]
```diff
@@ -389,11 +388,11 @@ genValueDefault w = genValue w genAssetId genSignedNonZeroQuantity
 
 -- | Generate a 'Value' suitable for minting, i.e. non-ADA asset ID and a
 -- positive or negative quantity.
-genValueForMinting :: MaryEraOnwards era -> Gen Value
+genValueForMinting :: forall era. MaryEraOnwards era -> Gen Value
 genValueForMinting w =
   fromLedgerValue sbe <$> genValue w genAssetIdNoAda genSignedNonZeroQuantity
  where
-  sbe = maryEraOnwardsToShelleyBasedEra w
+  sbe = inject w :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/gen/Test/Gen/Cardano/Api/Typed.hs:395]:
```suggestion
  sbe :: ShelleyBasedEra era = inject w
```

Rationale: It's more common to have the type attached to the variable

File: cardano-api/internal/Cardano/Api/Query/Expr.hs [hunk 2]
```diff
@@ -274,7 +275,7 @@ queryStakeDelegDeposits
 queryStakeDelegDeposits era stakeCreds
   | S.null stakeCreds = pure . pure $ pure mempty
   | otherwise = do
-      let sbe = babbageEraOnwardsToShelleyBasedEra era
+      let sbe = inject era :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Query/Expr.hs:278]:
```suggestion
      let sbe :: ShelleyBasedEra era = inject era
```

File: cardano-api/internal/Cardano/Api/Query/Expr.hs [hunk 3]
```diff
@@ -398,12 +399,12 @@ queryDRepState
           (Either EraMismatch (Map (L.Credential L.DRepRole L.StandardCrypto) (L.DRepState L.StandardCrypto)))
       )
 queryDRepState era drepCreds = do
-  let sbe = conwayEraOnwardsToShelleyBasedEra era
+  let sbe = inject era :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Query/Expr.hs:402]:
```suggestion
  let sbe :: ShelleyBasedEra era = inject era
```

File: cardano-api/internal/Cardano/Api/Query/Expr.hs [hunk 4]
```diff
@@ -472,12 +473,12 @@ queryStakeVoteDelegatees
           (Either EraMismatch (Map StakeCredential (L.DRep L.StandardCrypto)))
       )
 queryStakeVoteDelegatees era stakeCredentials = do
-  let sbe = conwayEraOnwardsToShelleyBasedEra era
+  let sbe = inject era :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Query/Expr.hs:476]:
```suggestion
  let sbe :: ShelleyBasedEra era = inject era
```

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 5]
```diff
@@ -2049,12 +2048,11 @@ fromAlonzoTxOut w txOut =
       (fromAlonzoTxOutDatumHash w (txOut ^. L.dataHashTxOutL))
       ReferenceScriptNone
  where
-  sbe = alonzoEraOnwardsToShelleyBasedEra w
+  sbe = inject w :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:2051]:
```suggestion
  sbe :: ShelleyBasedEra era = inject w
```

File: cardano-api/internal/Cardano/Api/Tx/Body.hs [hunk 6]
```diff
@@ -2069,7 +2067,7 @@ fromBabbageTxOut w txdatums txout =
           SJust rScript -> fromShelleyScriptToReferenceScript shelleyBasedEra rScript
       )
  where
-  sbe = babbageEraOnwardsToShelleyBasedEra w
+  sbe = inject w :: ShelleyBasedEra era
```
Comment (smelc) [inline cardano-api/internal/Cardano/Api/Tx/Body.hs:2070]:
```suggestion
  sbe :: ShelleyBasedEra era = inject w
```

**Review by smelc — APPROVED** [review]
Nice QoL change :+1:

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Please explain the motivation behind this change. I don't see how this helps us. We are adding an additional step to the deprecation of these era conversion functions where what we should be doing is figuring out how to change the interface then deprecating and eventually remove the functions.

**Comment by carbolymer** [comment]
@Jimbo4350 the goal of this PR is to make the codebase more pleasant to work with until we get rid of Eons. There's still some amount of work left to do, before we arrive there.

 I disagree that it adds an additional step in deprecation, I see `Inject` as a general purpose tool for total conversions between any two types. Those `Inject` instances would be gone together with `Eons` as well.

If you think this PR is changing too much, what do you think of just using `inject` in the glue code for experimental API. I mean the changes in `cardano-api/internal/Cardano/Api/Experimental/Eras.hs` ?

**Review by Jimbo4350 — APPROVED** [review]

**Pull request merged**